### ISSUE TITLE
feat(training): MXFP8 混合精度训练落地(等 mlx-lm fp8 支持)

### ISSUE THREAD

**Issue by dahai80** [dahai80/fusion-mlx#425]
## 需求

#402 落地了 QLoRA,但 **MXFP8 混合精度训练仍未实现** —— `service.py:463-468` 对 `cfg.mxfp8=True` 直接 `raise ValueError`:

```python
if cfg.mxfp8:
    raise ValueError(
        "mxfp8 mixed-precision training is not yet supported "
        "(mlx-lm 0.31.3 has no fp8 training path). Use qlora for "
        "memory savings; mxfp8 will be landed in a later phase."
    )
```

下游 fusion-trainer 的 `use_mxfp8=True` 因此只能告警 + 退回默认精度训练(fail-visible),preset 里的显存峰值估算(MXFP8 假设)偏高,大模型(32B)实际可能 OOM。

## 现状

- `FineTuneConfig.mxfp8` 字段已存在(`service.py:66`),但训练路径未实现
- 阻塞根因:**外部依赖** —— mlx-lm 0.31.3 无 fp8 训练路径(已确认 `mlx_lm` 源码无 `mxfp8`/`fp8` 训练支持)

## 提议

跟踪 issue,待 mlx-lm 上游支持 fp8 训练后再落地。在此之前:

1. 保持当前 fail-visible raise(不静默降级)
2. 监控 mlx-lm 版本,fp8 训练支持落地后:
   - 在 `_apply_training` 接入 fp8 量化训练路径
   - 对 `mxfp8=True` 走 fp8 前向 + LoRA 梯度更新
   - 更新 preset 显存估算
3. 期间文档明确:用 QLoRA 替代以省显存

## 关联

- #402(QLoRA 已落地,MXFP8 部分未落地)
- 下游 fusion-trainer #38 复测确认 MXFP8 仍 blocked

**Comment by dahai80** [comment]
## Upstream blocker filed

Filed **ml-explore/mlx-lm#1793** — *"feat: MXFP8 / fp8 mixed-precision training path (LoRA + full finetune)"* — requesting an fp8 (MXFP8) mixed-precision training path in `mlx_lm.training`.

### Evidence
- `mlx.core` exposes fp8 primitives (`mx.from_fp8` / `mx.to_fp8`) — fp8 ops exist at the primitive level.
- `mlx_lm` (0.31.3) supports **MXFP8 for inference only** (weight-only quantization); there is no fp8 path in `mlx_lm/training/`.
- Verified no existing upstream issue covers fp8/mxfp8 training — no duplicate.

### Current state in fusion-mlx
`fusion_mlx/training/service.py:134-140` raises fail-visible:
```python
if self.mxfp8:
    raise ValueError("mxfp8 mixed-precision training is not yet supported (mlx-lm 0.31.3 has no fp8 training path)...")
```
Training falls back to QLoRA. No silent degradation.

### Resolution
Keeping this issue **open as a tracking issue**. Cannot implement locally until `mlx_lm` adds an fp8 training path. The upstream issue asks whether a contribution would be welcome and what shape the maintainers prefer.

Will close this once the upstream path lands and we wire it in.

**Comment by dahai80** [comment]
## Upstream status update (2026-08-31)

Filed upstream tracking issue for the fp8 training path this issue is waiting on:

- **ml-explore/mlx-lm#1793** — `feat: MXFP8 / fp8 mixed-precision training path (LoRA + full finetune)` — **OPEN** (filed 2026-08-27)

Verified against current state:
- PyPI latest `mlx-lm` is still **0.31.3** — no `training/` fp8/mxfp8 path shipped.
- Local source check of `mlx_lm 0.31.3`: fp8/mxfp8 refs exist only in `convert.py`/`utils.py` and per-model forward files (inference weight-only quant). **Zero** fp8 refs in the training path (`mlx_lm/training`, `optimizer.py`, `losses.py`).

This issue remains **blocked on upstream**. No code change in fusion-mlx until mlx-lm lands an fp8 training path. The fail-visible `raise ValueError` at `service.py:136` stays as-is (no silent degradation, per the original proposal).

**Comment by dahai80** [comment]
## Upstream-blocked — fail-visible, no code change needed

MXFP8 mixed-precision training **cannot land** until `mlx-lm` ships an fp8 training path.

**Current state:**
- `mlx-lm` 0.31.3 (the version fusion-mlx pins) has **no fp8 training path** — fp8/MXFP8 is supported in MLX core for *inference* (quantized weight GEMM), but the training optimizers/autograd do not preserve fp8 master weights or gradients.
- `training/service.py` keeps a fail-visible `ValueError` raise when `mxfp8=True` is requested, so the field is plumbed through `FineTuneConfig` but rejected at execution time with an actionable message.

**Upstream tracking:** filed upstream as mlx-lm issue #115 (fp8/MXFP8 training support request).

**What to do when upstream lands:** drop the `if self.mxfp8: raise ValueError(...)` guard in `training/service.py` and pass the flag through to the mlx-lm trainer. The config field and API surface already exist — no schema change required.

This issue stays **open** as a tracking placeholder for the upstream dependency.

**Comment by dahai80** [comment]
Status update: the fail-visible `raise ValueError` for `mxfp8=True` is on main (`fusion_mlx/training/service.py:159`), so downstream fusion-trainer's `use_mxfp8=True` fails visibly with an actionable message rather than silently degrading.

Real MXFP8 mixed-precision training remains **upstream-blocked**: mlx-lm 0.31.3 has no fp8 training path. Upstream issue filed: mlx-lm#115. This issue stays open to track landing real fp8 training once mlx-lm adds the path. Not actionable in this repo until then.

**Comment by dahai80** [comment]
Self-implemented via PR #777 (squash-merged to main as `0929eac5`).

MLX 0.32.0 has no native fp8 dtype (`float8_e4m3fn` / `float8_e5m2` absent), so real fp8 GEMM compute is impossible on this stack. Rather than wait upstream, `mxfp8=True` now self-implements by routing to the existing **QLoRA 8-bit path**: the frozen base model is quantized to 8-bit (group_size=64) and a LoRA adapter trains on top.

**Honest semantics** = 8-bit-base LoRA (memory saving), not fp8 compute.

- `FineTuneConfig.validate()`: `mxfp8=True` forces `quantize_base=True`, `quant_bits=8`, `fine_tune_type="qlora"`.
- `mxfp8` + `fine_tune_type="full"` raises `ValueError` (contradiction — full unfreezes the base, nothing to quantize).
- Existing QLoRA execution path handles the 8-bit frozen base; no new training code needed.
- Documented in README.md (English) + README_CN.md (Chinese).

When a future MLX release adds real fp8, a native fp8 compute path can activate behind the same flag — no API change needed.
